puppet/modules/jenkins_job_builder/files/theforeman.org/pipelines/test/test_foreman.groovy

```diff
@@ -0,0 +1,199 @@
+pipeline {
+    agent { label 'fast' }
+
+    options {
+        timestamps()
+        timeout(time: 2, unit: 'HOURS')
+        ansiColor('xterm')
+    }
+
+    stages {
+        stage('Test Matrix') {
+            parallel {
+                stage('ruby-2.5-postgres') {
+                    agent { label 'fast' }
+                    environment {
+                        RUBY = '2.5'
+                        GEMSET = 'ruby-2.5-postgres'
+                    }
+                    stages {
+                        stage("setup-2.5-postgres") {
+                            steps {
+                                git url: 'https://github.com/theforeman/foreman', branch: 'develop'
+                                configureRVM(env.RUBY, env.GEMSET)
+                                databaseFile(gemset(env.GEMSET))
+                                configureDatabase(env.RUBY, env.GEMSET)
+                                withRVM(['npm install'], env.RUBY, env.GEMSET)
+                            }
+                        }
+                        stage("unit-tests-2.5-postgres") {
+                            steps {
+                                withRVM(['bundle exec rake jenkins:unit TESTOPTS="-v" --trace'], env.RUBY, env.GEMSET)
+                            }
+                        }
+                        stage("integration-tests-2.5-postgres") {
+                            steps {
+                                withRVM(['bundle exec rake jenkins:integration TESTOPTS="-v" --trace'], env.RUBY, env.GEMSET)
+                            }
+                        }
+                        stage("assets-precompile-2.5-postgres") {
+                            steps {
+                                withRVM(['bundle exec rake assets:precompile RAILS_ENV=production'], env.RUBY, env.GEMSET)
+                            }
+                        }
+                    }
+                    post {
```

Comment:
Does this mean we generate the source if this particular stage (ruby-2.5-postgres) passes? In the past we only built it if *all* stages* passed. Though I guess this is also covered by not continuing to the packaging stage.

Reply:
That is correct.  I was trying to avoid having to create a brand new environment just to generate the source. And the current job generates the source for *all* job configurations (https://github.com/theforeman/foreman-infra/blob/master/puppet/modules/jenkins_job_builder/files/theforeman.org/scripts/test/test_develop.sh#L44) but it only *archives* the source if all passed. However, workspaces play a role here. As you said, if these stages fail, the packaging won't kickoff.

Although, in this current design you did get me thinking that this may not work. As the packaging jobs are kicked off mid-pipeline and those packaging jobs need lastSuccessfulBuild which could have a race condition in current form. In @komidore64 proposal, a new packaging job would generate the source and avoid this. So perhaps we need to get that job online first.

Reply:
Looking at https://github.com/theforeman/foreman-packaging/blob/9293c65d55c59eec7a1a58b762f98ca48d12bb2f/rel-eng/custom/custom.py#L124-L130 I also wonder if this would break an assumption. Perhaps not because runs might not even show up in a pipeline build.